packages/code-block/partials/code-lines/index.tsx

```diff
@@ -0,0 +1,240 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+/**
+ * Note: lines of code are expected to be stable. If we need to work with
+ * dynamic code blocks in the future, we could assign random unique IDs
+ * to each line during the `linesOfCode` `useMemo` function.
+ *
+ * For now, we disable react/no-array-index key for the entire file.
+ */
+/* eslint-disable react/no-array-index-key */
+
+import React, { PropsWithChildren, ReactElement, useMemo } from 'react'
+import classNames from 'classnames'
+import parseHighlightedLines from '../../utils/parse-highlighted-lines'
+import splitJsxIntoLines from './utils/split-jsx-into-lines'
+import splitHtmlIntoLines from './utils/split-html-into-lines'
+import s from './code-lines.module.css'
+
+interface CodeLinesProps {
+  code: string | ReactElement
+  language?: string
+  lineNumbers?: boolean
+  highlight?: string
+  hasFloatingCopyButton?: boolean
+  wrapCode?: boolean
+}
+
+/**
+ * Render the provided code into separate line elements,
+ * accounting for all provided options.
+ */
+function CodeLines({
+  code,
+  language,
+  lineNumbers,
+  highlight,
+  hasFloatingCopyButton,
+  wrapCode,
+}: CodeLinesProps) {
+  // Parse out an array of integers representing which lines to highlight
+  const highlightedLines = parseHighlightedLines(highlight) as number[]
+
+  /**
+   * Split the incoming code into lines.
+   * We need to do this in order to render each line of code in a
+   * separate element, which is necessary for features such as highlighting
+   * specific lines and allowing code to wrap.
+   */
+  const linesOfCode = useMemo(() => {
+    const isHtmlString = typeof code === 'string'
+    const lineElements = isHtmlString
+      ? splitHtmlIntoLines(code)
+      : splitJsxIntoLines(code)
+    const lineCount = lineElements.length
+    const padLevel = Math.max(lineCount.toString().length, 1)
+    return lineElements.map((children, index) => {
+      const number = index + 1
+      const numberPadded = number.toString().padEnd(padLevel)
+      const highlight = highlightedLines.indexOf(number) !== -1
+      const dim = highlightedLines.length > 0 && !highlight
+      return { children, number: numberPadded, highlight, dim }
+    })
+  }, [code, highlightedLines])
+
+  // When rendering wrapped code with line numbers shown, we need a spacer value
+  // that matches the padding inset of all other line numbers
+  let numberSpacer: string | null = null
+  if (lineNumbers) {
+    const padLevel = Math.max(linesOfCode.length.toString().length, 1)
+    numberSpacer = ''.padEnd(padLevel)
+  }
+
+  // When the floating copy button is present, we add padding to many lines
+  const padRight = Boolean(hasFloatingCopyButton)
+
+  if (wrapCode) {
```

Comment:
I've refactored this component fairly significantly, which felt necessary in order to make the `wrapCode` option feasible.

For some context, there are fundamentally different layout needs with `wrapCode` compared to without:
- When allowing overflow, we want the lines of code to scroll together, but we want the line numbers for each line to stay in a fixed position, like a frozen spreadsheet column. This necessitates a two-column layout at the top level, with the line numbers in one column, and the line contents in a separate adjacent column with scrollable content.
- When wrapping code, a single line will take up more vertical space when it wraps. The two-column layout needed for overflowing code is no longer usable - the only approach I can imagine would be to somehow detect wrapping in each line in the "lines" columns and constantly update the corresponding line number contents in the "numbers" column to contain an identical number of newlines. It seemed much easier to rethink this use case as an alternate layout of the same atomic components, with a single column where each "row" is a flex layout that contains both the line number and the line contents.

To achieve these varied layouts, the previous `CodeLines` component was first refactored into smaller composable components, and then here in this conditional, we compose those smaller components into the appropriate layouts depending on the incoming `wrapCode` boolean.